File: app/src/main/java/app/kamy/saatApp/domain/prayer/PrayerCalculationMethod.kt
package app.kamy.saatApp.domain.prayer

enum class PrayerCalculationMethod(val rawValue: String) {
    MUHAMMADIYAH("muhammadiyah"),
    KEMENAG("kemenag"),
    MUIS("muis"),
    JAKIM("jakim"),
    BRUNEI("brunei"),
    KARACHI("karachi"),
    TEHRAN("tehran"),
    JAFARI("jafari"),
    ISNA("isna"),
    MWL("mwl"),
    UMM_AL_QURA("ummAlQura"),
    EGYPTIAN("egyptian"),
    MCW("mcw"),
    GULF("gulf"),
    KUWAIT("kuwait"),
    QATAR("qatar"),
    DUBAI("dubai"),
    TUNISIA("tunisia"),
    ALGERIA("algeria"),
    MOROCCO("morocco"),
    JORDAN("jordan"),
    FRANCE("france"),
    TURKEY("turkey"),
    RUSSIA("russia"),
    LISBON("lisbon");

    val displayName: String
        get() = when (this) {
            MUHAMMADIYAH -> "Muhammadiyah"
            KEMENAG -> "Kementerian Agama RI (Kemenag)"
            MUIS -> "Majlis Ugama Islam Singapura (MUIS)"
            JAKIM -> "Jabatan Kemajuan Islam Malaysia (JAKIM)"
            BRUNEI -> "Majlis Ugama Islam Brunei (MUIB)"
            KARACHI -> "University of Islamic Sciences, Karachi"
            TEHRAN -> "Institute of Geophysics, University of Tehran"
            JAFARI -> "Shia Ithna-Ashari (Leva Institute, Qom)"
            ISNA -> "Islamic Society of North America (ISNA)"
            MWL -> "Muslim World League (MWL)"
            UMM_AL_QURA -> "Umm Al-Qura University, Makkah"
            EGYPTIAN -> "Egyptian General Authority of Survey"
            MCW -> "Moonsighting Committee Worldwide"
            GULF -> "Gulf Region (90 min Isha)"
            KUWAIT -> "Ministry of Awqaf & Islamic Affairs, Kuwait"
            QATAR -> "Ministry of Awqaf & Islamic Affairs, Qatar"
            DUBAI -> "Islamic Affairs (IACAD), Dubai"
            TUNISIA -> "Ministère des Affaires Religieuses, Tunisie"
            ALGERIA -> "Ministère des Affaires Religieuses, Algérie"
            MOROCCO -> "Ministère des Habous et des Affaires Islamiques, Maroc"
            JORDAN -> "Ministry of Awqaf & Islamic Affairs, Jordan"
            FRANCE -> "Union des Organisations Islamiques de France (UOIF)"
            TURKEY -> "Diyanet İşleri Başkanlığı, Türkiye"
            RUSSIA -> "Spiritual Administration of Muslims of Russia (DUM RF)"
            LISBON -> "Comunidade Islâmica de Lisboa, Portugal"
        }

    val organization: String
        get() = when (this) {
            MUHAMMADIYAH -> "Persyarikatan Muhammadiyah"
            KEMENAG -> "Kementerian Agama Republik Indonesia"
            MUIS -> "Majlis Ugama Islam Singapura"
            JAKIM -> "Jabatan Kemajuan Islam Malaysia"
            BRUNEI -> "Kementerian Hal Ehwal Ugama, Brunei Darussalam"
            KARACHI -> "Jamia Uloom-ul-Islamia, Karachi, Pakistan"
            TEHRAN -> "Institute of Geophysics, University of Tehran, Iran"
            JAFARI -> "Leva Research Institute, Qom, Iran"
            ISNA -> "Islamic Society of North America"
            MWL -> "Muslim World League (Rabitah al-Alam al-Islami)"
            UMM_AL_QURA -> "Umm Al-Qura University, Makkah al-Mukarramah"
            EGYPTIAN -> "Egyptian General Authority of Survey (Al-Hai'ah Al-Misriyyah)"
            MCW -> "Moonsighting Committee Worldwide"
            GULF -> "Gulf Region Prayer Calculation (90 min interval)"
            KUWAIT -> "Ministry of Awqaf and Islamic Affairs, Kuwait"
            QATAR -> "Ministry of Awqaf and Islamic Affairs, Qatar"
            DUBAI -> "Islamic Affairs & Charitable Activities Department, Dubai"
            TUNISIA -> "Ministère des Affaires Religieuses, République Tunisienne"
            ALGERIA -> "Ministère des Affaires Religieuses et des Wakfs, Algérie"
            MOROCCO -> "Ministère des Habous et des Affaires Islamiques, Royaume du Maroco"
            JORDAN -> "Ministry of Awqaf, Islamic Affairs and Holy Places, Jordan"
            FRANCE -> "Musulmans de France (ex-UOIF)"
            TURKEY -> "T.C. Diyanet İşleri Başkanlığı"
            RUSSIA -> "Spiritual Administration of Muslims of the Russian Federation"
            LISBON -> "Comunidade Islâmica de Lisboa, Portugal"
        }

    @get:androidx.annotation.DrawableRes
    val iconRes: Int
        get() = when (this) {
            MUHAMMADIYAH -> app.kamy.saatApp.R.drawable.institution_muhammadiyah
            KEMENAG -> app.kamy.saatApp.R.drawable.institution_kemenag
            MUIS -> app.kamy.saatApp.R.drawable.institution_muis
            JAKIM -> app.kamy.saatApp.R.drawable.institution_jakim
            BRUNEI -> app.kamy.saatApp.R.drawable.institution_muib
            KARACHI -> app.kamy.saatApp.R.drawable.institution_karachi
            TEHRAN -> app.kamy.saatApp.R.drawable.institution_tehran
            JAFARI -> app.kamy.saatApp.R.drawable.institution_jafari
            ISNA -> app.kamy.saatApp.R.drawable.institution_isna
            MWL -> app.kamy.saatApp.R.drawable.institution_mwl
            UMM_AL_QURA -> app.kamy.saatApp.R.drawable.institution_umm_al_qura
            EGYPTIAN -> app.kamy.saatApp.R.drawable.institution_egyptian
            MCW -> app.kamy.saatApp.R.drawable.institution_mcw
            GULF -> app.kamy.saatApp.R.drawable.institution_gulf
            KUWAIT -> app.kamy.saatApp.R.drawable.institution_kuwait
            QATAR -> app.kamy.saatApp.R.drawable.institution_qatar
            DUBAI -> app.kamy.saatApp.R.drawable.institution_dubai
            TUNISIA -> app.kamy.saatApp.R.drawable.institution_tunisia
            ALGERIA -> app.kamy.saatApp.R.drawable.institution_algeria
            MOROCCO -> app.kamy.saatApp.R.drawable.institution_morocco
            JORDAN -> app.kamy.saatApp.R.drawable.institution_jordan
            FRANCE -> app.kamy.saatApp.R.drawable.institution_france
            TURKEY -> app.kamy.saatApp.R.drawable.institution_turkey
            RUSSIA -> app.kamy.saatApp.R.drawable.institution_russia
            LISBON -> app.kamy.saatApp.R.drawable.institution_lisbon
        }

    val aladhanMethodId: Int
        get() = when (this) {
            JAFARI -> 0
            KARACHI -> 1
            ISNA -> 2
            MWL -> 3
            UMM_AL_QURA -> 4
            EGYPTIAN -> 5
            TEHRAN -> 7
            GULF -> 8
            KUWAIT -> 9
            QATAR -> 10
            MUIS -> 11
            FRANCE -> 12
            TURKEY -> 13
            RUSSIA -> 14
            MCW -> 15
            DUBAI -> 16
            JAKIM, BRUNEI -> 17
            TUNISIA -> 18
            ALGERIA -> 19
            KEMENAG -> 20
            MOROCCO -> 21
            LISBON -> 22
            JORDAN -> 23
            MUHAMMADIYAH -> 99
        }

    val aladhanMethodSettings: String?
        get() = when (this) {
            MUHAMMADIYAH -> "18,null,18"
            KEMENAG, MUIS, JAKIM, BRUNEI, JORDAN -> "20,null,18"
            KARACHI -> "18,null,18"
            ISNA -> "15,null,15"
            MWL -> "18,null,17"
            MCW -> "18,null,18"
            EGYPTIAN -> "19.5,null,17.5"
            UMM_AL_QURA -> "18.5,null,90 min"
            KUWAIT -> "18,null,17.5"
            QATAR -> "18,null,90 min"
            DUBAI -> "18.2,null,18.2"
            TUNISIA -> "18,null,18"
            ALGERIA -> "18,null,17"
            MOROCCO -> "19,null,17"
            FRANCE, LISBON -> "12,null,12"
            TURKEY -> "18,null,17"
            RUSSIA -> "16,null,15"
            TEHRAN -> "17.7,4.5,14"
            JAFARI -> "16,4,14"
            GULF -> "19.5,null,90 min"
        }

    val countryName: String
        get() = when (this) {
            MUHAMMADIYAH, KEMENAG -> "Indonesia"
            MUIS -> "Singapura (Singapore)"
            JAKIM -> "Malaysia"
            BRUNEI -> "Brunei Darussalam"
            KARACHI -> "Pakistan / India / Bangladesh"
            TEHRAN, JAFARI -> "Iran / Shia"
            ISNA -> "Amerika Utara (USA / Canada)"
            MWL, MCW -> "Internasional (Worldwide)"
            UMM_AL_QURA -> "Arab Saudi (Saudi Arabia)"
            EGYPTIAN -> "Mesir (Egypt)"
            GULF -> "Kawasan Teluk (Gulf Region)"
            KUWAIT -> "Kuwait"
            QATAR -> "Qatar"
            DUBAI -> "Uni Emirat Arab (UAE - Dubai)"
            TUNISIA -> "Tunisia"
            ALGERIA -> "Aljazair (Algeria)"
            MOROCCO -> "Maroko (Morocco)"
            JORDAN -> "Yordania (Jordan)"
            FRANCE -> "Perancis (France)"
            TURKEY -> "Turki (Turkey)"
            RUSSIA -> "Rusia (Russia)"
            LISBON -> "Portugal (Lisbon)"
        }

    val aladhanSchool: Int
        get() = when (this) {
            KARACHI, TURKEY, RUSSIA -> 1 // 1 = Hanafi (bayangan Ashar 2x)
            else -> 0 // 0 = Standar (Syafi'i, Maliki, Hanbali / bayangan Ashar 1x)
        }

    val aladhanTune: String
        get() = when (this) {
            MUHAMMADIYAH -> "0,2,-1,1,1,3,0,2,0"
            KEMENAG -> "0,0,-1,1,1,3,0,2,0"
            MUIS, JAKIM, BRUNEI -> "0,0,0,0,0,0,0,0,0"
            else -> "0,0,0,0,0,0,0,0,0"
        }

    companion object {
        val defaultMethod: PrayerCalculationMethod = KEMENAG

        fun fromRawValue(raw: String?): PrayerCalculationMethod =
            entries.firstOrNull { it.rawValue == raw } ?: defaultMethod

        fun fromAladhanId(id: Int): PrayerCalculationMethod =
            entries.firstOrNull { it.aladhanMethodId == id } ?: defaultMethod

        fun forCountryCode(code: String): PrayerCalculationMethod = when (code.uppercase()) {
            "ID" -> KEMENAG
            "SG" -> MUIS
            "MY" -> JAKIM
            "BN" -> BRUNEI
            "PK", "IN", "BD" -> KARACHI
            "SA" -> UMM_AL_QURA
            "EG" -> EGYPTIAN
            "TR" -> TURKEY
            "FR" -> FRANCE
            "US", "CA" -> ISNA
            "AE" -> DUBAI
            "KW" -> KUWAIT
            "QA" -> QATAR
            "MA" -> MOROCCO
            "DZ" -> ALGERIA
            "TN" -> TUNISIA
            "JO" -> JORDAN
            "IR" -> TEHRAN
            "RU" -> RUSSIA
            "PT" -> LISBON
            else -> MWL
        }
    }
}

File: iosApp/Saat/Core/Prayer/PrayerCalculationMethod.swift
//
//  PrayerCalculationMethod.swift
//  Sāat
//
//  Created by Elmee on 25/04/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

enum PrayerCalculationMethod: String, CaseIterable, Sendable, Identifiable, Codable {
    // Southeast Asia (featured)
    case muhammadiyah
    case kemenag
    case muis
    case jakim
    case brunei
    case karachi
    case tehran
    case jafari
    case isna
    case mwl
    case ummAlQura
    case egyptian
    case mcw
    case gulf
    case kuwait
    case qatar
    case dubai
    case tunisia
    case algeria
    case morocco
    case jordan
    case france
    case turkey
    case russia
    case lisbon

    static let storageKey = "prayer_calculation_method"
    static let defaultMethod: PrayerCalculationMethod = .muhammadiyah

    var id: String { rawValue }

    enum Section: String, CaseIterable, Identifiable, Sendable {
        case southeastAsia = "Southeast Asia"
        case southAsia = "South Asia"
        case global = "Global standards"
        case middleEast = "Middle East & North Africa"
        case europe = "Europe"

        var id: String { rawValue }

        var methods: [PrayerCalculationMethod] {
            PrayerCalculationMethod.allCases.filter { $0.section == self }
        }
    }

    var section: Section {
        switch self {
        case .muhammadiyah, .kemenag, .muis, .jakim, .brunei:
            return .southeastAsia
        case .karachi, .tehran:
            return .southAsia
        case .jafari, .isna, .mwl, .ummAlQura, .egyptian, .mcw:
            return .global
        case .gulf, .kuwait, .qatar, .dubai, .tunisia, .algeria, .morocco, .jordan:
            return .middleEast
        case .france, .turkey, .russia, .lisbon:
            return .europe
        }
    }

    var displayName: String {
        switch self {
        case .muhammadiyah: "Muhammadiyah"
        case .kemenag: "Ministry of Religious Affairs (Kemenag)"
        case .muis: "Majlis Ugama Islam Singapura (MUIS)"
        case .jakim: "Jabatan Kemajuan Islam Malaysia (JAKIM)"
        case .brunei: "Majlis Ugama Islam Brunei (MUIB)"
        case .karachi: "Karachi"
        case .tehran: "Tehran"
        case .jafari: "Jafari"
        case .isna: "ISNA"
        case .mwl: "Muslim World League"
        case .ummAlQura: "Umm Al-Qura"
        case .egyptian: "Egyptian"
        case .mcw: "Moonsighting Committee"
        case .gulf: "Gulf Region"
        case .kuwait: "Kuwait"
        case .qatar: "Qatar"
        case .dubai: "Dubai"
        case .tunisia: "Tunisia"
        case .algeria: "Algeria"
        case .morocco: "Morocco"
        case .jordan: "Jordan"
        case .france: "UOIF France"
        case .turkey: "Diyanet Turkey"
        case .russia: "Russia"
        case .lisbon: "Lisbon"
        }
    }

    var organization: String {
        switch self {
        case .muhammadiyah: "Persyarikatan Muhammadiyah"
        case .kemenag: "Ministry of Religious Affairs (Kemenag)"
        case .muis: "Majlis Ugama Islam Singapura"
        case .jakim: "Jabatan Kemajuan Islam Malaysia"
        case .brunei: "Majlis Ugama Islam Brunei"
        case .karachi: "University of Islamic Sciences, Karachi"
        case .tehran: "Institute of Geophysics, University of Tehran"
        case .jafari: "Shia Ithna-Ashari (Leva Institute, Qom)"
        case .isna: "Islamic Society of North America"
        case .mwl: "Muslim World League"
        case .ummAlQura: "Umm Al-Qura University, Makkah"
        case .egyptian: "Egyptian General Authority of Survey"
        case .mcw: "Moonsighting Committee Worldwide"
        case .gulf: "Gulf Region"
        case .kuwait: "Kuwait"
        case .qatar: "Qatar"
        case .dubai: "Dubai (experimental)"
        case .tunisia: "Tunisia"
        case .algeria: "Algeria"
        case .morocco: "Morocco"
        case .jordan: "Ministry of Awqaf, Jordan"
        case .france: "Union Organization islamic de France"
        case .turkey: "Diyanet İşleri Başkanlığı"
        case .russia: "Spiritual Administration of Muslims of Russia"
        case .lisbon: "Comunidade Islamica de Lisboa"
        }
    }

    var region: String {
        switch section {
        case .southeastAsia: "Southeast Asia"
        case .southAsia: "South Asia"
        case .global: "Worldwide"
        case .middleEast: "MENA & Gulf"
        case .europe: "Europe"
        }
    }

    var subtitle: String {
        "\(displayName) · \(organization)"
    }

    var pickerSubtitle: String {
        organization
    }

    static var hasSavedPreference: Bool {
        UserDefaults.standard.string(forKey: storageKey) != nil
    }

    static func savedOrDefault() -> PrayerCalculationMethod {
        guard let raw = UserDefaults.standard.string(forKey: storageKey),
              let method = PrayerCalculationMethod(rawValue: raw) else {
            return defaultMethod
        }
        return method
    }

    func persist(notify: Bool = true) {
        UserDefaults.standard.set(rawValue, forKey: Self.storageKey)
        if notify {
            NotificationCenter.default.post(name: .prayerCalculationMethodDidChange, object: nil)
        }
    }

    static func forCountryCode(_ code: String) -> PrayerCalculationMethod {
        switch code.uppercased() {
        case "ID": return .muhammadiyah
        case "SG": return .muis
        case "MY": return .jakim
        case "BN": return .brunei
        case "PK", "IN", "BD": return .karachi
        case "SA": return .ummAlQura
        case "EG": return .egyptian
        case "TR": return .turkey
        case "FR": return .france
        case "US", "CA": return .isna
        case "AE": return .dubai
        case "KW": return .kuwait
        case "QA": return .qatar
        case "MA": return .morocco
        case "DZ": return .algeria
        case "TN": return .tunisia
        case "JO": return .jordan
        case "IR": return .tehran
        case "RU": return .russia
        case "PT": return .lisbon
        default: return .mwl
        }
    }

    var aladhanMethodID: Int {
        switch self {
        case .jafari: 0
        case .karachi: 1
        case .isna: 2
        case .mwl: 3
        case .ummAlQura: 4
        case .egyptian: 5
        case .tehran: 7
        case .gulf: 8
        case .kuwait: 9
        case .qatar: 10
        case .muis: 11
        case .france: 12
        case .turkey: 13
        case .russia: 14
        case .mcw: 15
        case .dubai: 16
        case .jakim, .brunei: 17
        case .tunisia: 18
        case .algeria: 19
        case .kemenag: 20
        case .morocco: 21
        case .lisbon: 22
        case .jordan: 23
        case .muhammadiyah: 99
        }
    }

    var aladhanMethodSettings: String? {
        switch self {
        case .muhammadiyah: "18,null,18"
        default: nil
        }
    }

    var aladhanSchool: Int { 0 }

    var aladhanTune: String {
        switch self {
        case .muhammadiyah: "0,2,-1,1,1,3,0,2,0"
        case .kemenag: "0,0,-1,1,1,3,0,2,0"
        case .muis, .jakim, .brunei: "0,0,0,0,0,0,0,0,0"
        default: "0,0,0,0,0,0,0,0,0"
        }
    }
}

extension Notification.Name {
    static let prayerCalculationMethodDidChange = Notification.Name("prayerCalculationMethodDidChange")
}
